fsspec/utils.py

```diff
@@ -483,3 +483,40 @@ def wrapper(cls):
 @contextmanager
 def nullcontext(obj):
     yield obj
+
+
+def merge_offset_ranges(paths, starts, ends, max_gap=0, max_block=None, **kwargs):
+    """Merge adjacent byte-offset ranges when the inter-range
+    gap is <= `max_gap`, and when the merged byte range does not
+    exceed `max_block` (if specified). This utility assumes that
+    the byte-offset ranges are ordered.
+    """
+
+    if len(starts) != len(paths) or len(ends) != len(paths):
+        raise ValueError
+
+    if paths:
+        # Loop through the coupled `paths`, `starts`, and
+        # `ends`, and merge adjacent blocks when appropriate
+        new_paths = paths[:1]
+        new_starts = starts[:1]
+        new_ends = ends[:1]
+        for i in range(1, len(paths)):
+            if (
+                paths[i] != paths[i - 1]
```

Comment:
There is an implicit requirement here that the paths be ordered and that the offsets per path also be ordered. That should at least be documented.

Reply:
Good point - I added an explicit sorting step to `merge_offset_ranges`, which the user can skip with `sort=False` if they know their input is already ordered.